curp/src/client.rs

```diff
@@ -87,16 +129,36 @@ where
                     continue;
                 }
             };
-            let term_valid = match resp.term() {
-                t if t > max_term => {
-                    // state reset
-                    ok_cnt = 0;
-                    max_term = resp.term();
-                    execute_result = None;
-                    true
+            let term_valid = {
+                let mut state = self.state.write();
+                match state.term.cmp(&resp.term()) {
+                    Ordering::Less => {
+                        // reset term only when the resp has leader id to prevent:
+                        // If a server loses contact with its leader, it will update its term for election. Since other servers are all right, the election will not succeed.
+                        // But if the client learns about the new term and updates its term to it, it will never get the true leader.
+                        if let Some(ref leader_id) = resp.leader_id {
+                            ok_cnt = 0;
+                            state.update_to_term(resp.term());
+                            state.set_leader(leader_id.clone());
+                            execute_result = None;
+                        }
+                        true
```

Comment:
I guess we can't count it as valid term unless the leader_id is determined.

Reply:
Discussed in Feishu.